src/connectors/cloudtrail.py

```diff
@@ -0,0 +1,198 @@
+    , value:userIdentity.sessionContext.sessionIssuer.arn::STRING user_identity_session_context_session_issuer_arn
+    , value:userIdentity.sessionContext.sessionIssuer.accountId::STRING user_identity_session_context_session_issuer_account_id
+    , value:userIdentity.sessionContext.sessionIssuer.userName::STRING user_identity_session_context_session_issuer_user_name
+    , value:errorCode::STRING error_code
+    , value:errorMessage::STRING error_message
+    , value:additionalEventData::VARIANT additional_event_data
+    , value:apiVersion::STRING api_version
+    , value:readOnly::BOOLEAN read_only
+    , value:resources::VARIANT resources
+    , value:serviceEventDetails::STRING service_event_details
+    , value:sharedEventID::STRING shared_event_id
+    , value:vpcEndpointID::STRING vpc_endpoint_id
+FROM DATA.{name}_STREAM, table(flatten(input => v:Records))
+WHERE ARRAY_SIZE(v:Records) > 0
+"""
+
+    comment = f"""
+---
+name: {name}
+"""
+
+    results = {}
+
+    try:
+        db.create_stage(name=name+'_STAGE', url=bucket, prefix=prefix, cloud='aws',
+                        credentials=role, file_format=FILE_FORMAT)
+        results['stage'] = 'success'
+    except Exception as e:
+        results['stage'] = 'failure'
+        results['exception'] = e
+        return results
+
+    try:
+        db.create_table(name=name+'_STAGING', cols='(v variant)')
+        results['staging_table'] = 'success'
+    except Exception as e:
+        results['staging_table'] = 'failure'
+        results['exception'] = e
+        return results
+
+    try:
+        db.create_stream(name=name+'_STREAM', target=name+'_STAGING')
+        results['stream'] = 'success'
+    except Exception as e:
+        results['stream'] = 'failure'
+        results['exception'] = e
+        return results
+
+    try:
+        pipe_sql = f"COPY INTO DATA.{name}_STAGING FROM @DATA.{name}_STAGE/"
+        db.create_pipe(name=name+'_PIPE', sql=pipe_sql, replace=True)
+        results['pipe'] = 'success'
+    except Exception as e:
+        results['pipe'] = 'failure'
+        results['exception'] = e
+        return results
+
+    try:
+        db.create_table(name=name+"_EVENTS_CONNECTION", cols=CLOUDTRAIL_LANDING_TABLE, comment=comment)
+        results['events_table'] = 'success'
+    except Exception as e:
+        results['events_table'] = 'failure'
+        results['exception'] = e
+        return results
+
+    try:
+        db.create_task(name=name+'_TASK', schedule='15 minutes',
+                       warehouse=WAREHOUSE, sql=cloudtrail_ingest_task)
+        results['task'] = 'success'
+    except Exception as e:
+        results['task'] = 'failure'
+        results['exception'] = e
+        return results
+
+    data = db.execute(f'DESC STAGE DATA.{name}_STAGE')
+    desc = list(data)
+    for i in desc:
+        if i[0] is 'STAGE_CREDENTIALS':
+            results.append(i)
+
+    return results
```

Comment:
I know I mentioned this as a way Go does it (instead of having exceptions) but seeing it implemented makes me realize this is really un-Pythonic. let's just raise exceptions through instead of returning errors.

Reply:
when successful, the return value here should be an updated version of the documentation in markdown. made [SP](https://snowflakecomputing.atlassian.net/browse/SP-1229) with details